File: CLASS3/2630.py
import sys
input = sys.stdin.readline

def check(x1, y1, x2, y2):
    num = board[y1][x1]
    if x1 == x2 and y1 == y2:
        return num
    for y in range(y1, y2+1):
        for x in range(x1, x2 + 1):
            if board[y][x] != num:
                return -1
    return num


def solve(x1, y1, x2, y2):
    global one_cnt
    global zero_cnt
    ck = check(x1, y1, x2, y2)
    if ck == 1:
        one_cnt +=1
        return
    if ck == 0:
        zero_cnt += 1
        return


    solve(x1, y1, (x1+x2)//2, (y1+y2)//2)
    solve((x1+x2)//2 +1, y1, x2, (y1+y2)//2)
    solve(x1, (y1+y2)//2 + 1, (x1+x2)//2, y2)
    solve((x1+x2)//2 +1, (y1+y2)//2 + 1, x2, y2)

n = int(input())
board = []
one_cnt =0
zero_cnt = 0

for _ in range(n):
    line = list(map(int, input().split()))
    board.append(line)

solve(0, 0, n-1, n-1)
print(zero_cnt)
print(one_cnt)

File: solved.ac/CLASS3_JS/2630.js
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const stdin = fs
  .readFileSync(inputPath)
  .toString()
  .split("\n")
  .map((line) => line.trim());

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const solution = () => {
  const n = +input();
  const board = Array.from({ length: n }, () => input().split(" ").map(Number));
  const ans = [0, 0];

  const check = (x1, y1, x2, y2, target) => {
    for (let y = y1; y < y2; y++) {
      for (let x = x1; x < x2; x++) {
        if (board[y][x] !== target) return false;
      }
    }
    return true;
  };

  const recur = (x1, y1, x2, y2, len) => {
    const target = board[y1][x1];
    if (check(x1, y1, x2, y2, target)) {
      ans[target] += 1;
      return;
    }

    const nextLen = (len / 2) | 0;
    for (let y = y1; y < y2; y += nextLen) {
      for (let x = x1; x < x2; x += nextLen) {
        recur(x, y, x + nextLen, y + nextLen, nextLen);
      }
    }
  };
  recur(0, 0, n, n, n);
  console.log(ans.join("\n"));
};

solution();
